src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,285 @@
+            .open(&self.file_path)
+            .map_err(Error::OpenFile)?;
+
+        let mut features = self.virtio_cfg.driver_features;
+        if self.read_only {
+            // Not sure if the driver is expected to explicitly acknowledge the `RO` feature,
+            // so adding it explicitly here when present just in case.
+            features |= 1 << VIRTIO_BLK_F_RO;
+        }
+
+        // TODO: Create the backend earlier (as part of `Block::new`)?
+        let disk = StdIoBackend::new(file, features).map_err(Error::Backend)?;
+
+        let driver_notify = SingleFdSignalQueue {
+            irqfd: self.irqfd.clone(),
+            interrupt_status: self.virtio_cfg.interrupt_status.clone(),
+        };
+
+        let inner = InOrderQueueHandler {
+            driver_notify,
+            queue: self.virtio_cfg.queues[0].clone(),
+            disk,
+        };
+
+        let handler = Arc::new(Mutex::new(QueueHandler { inner, ioeventfd }));
+
+        // Register the queue handler with the `EventManager`. We could record the `sub_id`
+        // (and/or keep a handler clone) for further interaction (i.e. to remove the subscriber at
+        // a later time, retrieve state, etc).
+        let _sub_id = self
+            .endpoint
+            .call_blocking(move |mgr| -> EvmgrResult<SubscriberId> {
+                Ok(mgr.add_subscriber(handler))
+            })
+            .map_err(Error::Endpoint)?;
+
+        self.virtio_cfg.device_activated = true;
+
+        Ok(())
+    }
+
+    fn reset(&mut self) -> Result<()> {
+        // Not implemented for now.
+        Ok(())
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> VirtioMmioDevice<M> for Block<M> {}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> MutDeviceMmio for Block<M> {
+    fn mmio_read(&mut self, _base: MmioAddress, offset: u64, data: &mut [u8]) {
+        self.read(offset, data);
+    }
+
+    fn mmio_write(&mut self, _base: MmioAddress, offset: u64, data: &[u8]) {
+        self.write(offset, data);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use vmm_sys_util::tempfile::TempFile;
+
+    use crate::virtio::tests::CommonArgsMock;
+
+    use super::*;
+
+    // Restricting this for now, because registering irqfds does not work on Arm without properly
+    // setting up the equivalent of the irqchip first (as part of `CommonArgsContext::new`).
+    #[cfg_attr(target_arch = "aarch64", ignore)]
+    #[test]
+    fn test_device() {
+        let mut ctx = CommonArgsMock::new();
+        let tmp = TempFile::new().unwrap();
+
+        let common_args = ctx.args();
+        let args = BlockArgs {
+            common: common_args,
+            file_path: tmp.as_path().to_path_buf(),
+            read_only: false,
+            root_device: true,
+        };
+
+        let _block = Block::new(args).unwrap();
+
+        // Some quick sanity checks. Most of the functionality around the device should be
+        // exercised/validated via integration tests.
```

Comment:
Hmm, can't we at least check that the device features from `virtio_cfg` are the expected ones and maybe check other scenarios as well (e.g. read_only -> true, root_device -> false)?